staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/interface.go

```diff
@@ -0,0 +1,98 @@
+// management filter makes a QueueSet for each priority level.  The
+// clock, and any other test-facilitating infrastructure, to use in
+// each QueueSet is known to the factory, the client does not need to
+// pass this stuff in each call to NewQueueSet.
+type QueueSetFactory interface {
+	NewQueueSet(name string, concurrencyLimit, numQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) (QueueSet, error)
+}
+
+// QueueSet is the abstraction for the queuing and dispatching
+// functionality of one non-exempt priority level.  It covers the
+// functionality described in the "Assignment to a Queue", "Queuing",
+// and "Dispatching" sections of
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+// .  Some day we may have connections between priority levels, but
+// today is not that day.
+type QueueSet interface {
+	// SetConfiguration updates the configuration
+	SetConfiguration(concurrencyLimit, desiredNumQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) error
+
+	// Quiesce controls whether this system is quiescing.  Passing a
+	// non-nil handler means the system should become quiescent, a nil
+	// handler means the system should become non-quiescent.  A call
+	// to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.  If all the queues have no requests
+	// waiting nor executing while the system is quiescent then the
+	// handler will eventually be called with no locks held (even if
+	// the system becomes non-quiescent between the triggering state
+	// and the required call).
+	//
+	// The filter uses this for a priority level that has become
+	// undesired, setting a handler that will cause the priority level
+	// to eventually be removed from the filter if the filter still
+	// wants that.  If the filter later changes its mind and wants to
+	// preserve the priority level then the filter can use this to
+	// cancel the handler registration.
+	Quiesce(EmptyHandler)
+
+	// Wait, in the happy case, shuffle shards the given request into
+	// a queue and eventually dispatches the request from that queue.
+	// Dispatching means to return with `quiescent==false` and
+	// `execute==true`.  In one unhappy case the request is
+	// immediately rebuffed with `quiescent==true` (which tells the
+	// filter that there has been a timing splinter and the filter
+	// re-calcuates the priority level to use); in all other cases
+	// `quiescent` will be returned `false` (even if the system is
+	// quiescent by then).  In the non-quiescent unhappy cases the
+	// request is eventually rejected, which means to return with
+	// `execute=false`.  In the happy case the caller is required to
+	// invoke the returned `afterExecution` after the request is done
+	// executing.  The hash value and hand size are used to do the
+	// shuffle sharding.
+	Wait(hashValue uint64) (quiescent, execute bool, afterExecution func())
```

Comment:
If I understand what this does, it needs to accept a context.Context so that it can abort early if the request is cancelled.

Reply:
It is my understanding that requests are cancelled only by the filter that imposes a service time limit.  We set the queuing time limit well under that, so cancellations are not expected to happen while a request is in Wait().

Reply:
In theory, the user behind the call can cancel in early, by e.g. closing the connection. This is supposed to trigger a context cancellation; we should free up resources as soon as possible when that happens. Other places in apiserver currently do it wrong, but we don't need to do it wrong here.

Also: I would expect it is possible for requests to sit in the queue for many seconds when apiserver is under load, I don't see why it would be significantly under the max request time, e.g. I'd expect it to be fine for a request to wait 59 out of the 60 seconds. Otherwise, apiserver has less scope to improve system performance by reordering requests.

Reply:
In the full implementation there is a requestmanagment filter which uses a requestmanager which uses this queueset implementation:
- [staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go](https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go#L68-L85)
- [staging/src/k8s.io/apiserver/pkg/util/flowcontrol/reqmgmt.go ](https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/reqmgmt.go#L234-L263)

The context-timeout handling is done in the requestmanagement filter.  Is the suggestion here to change this to have the context be wired through?

Reply:
@lavalamp : the code in the feature branch sets the limit on the amount of time that a request may wait in a queue to 25% of the max request time (the thing that defaults to 1 minute).  In the networking world the experience is that a lot of queuing is bad.  In our case it would seem a waste to allow a request that will take 10 seconds to execute wait for 59 seconds in a queue and then get aborted after it has run for only 1 second.

I can get behind plumbing the cancellation through for the sake of generality.  My first thought about how to do that would be to make QueueSet::Wait take a cancel channel as a parameter.  There is a standard way to extract that from a Context.  Or is there a reason to take a whole Context?

Reply:
Most (median) of our short-running (ordinary) requests have run times in the milliseconds last I saw.

We can also change the timeout behavior to permit X time in the queue and Y time for executing afterwards, tuning X and Y such that we expect requests will usually not wait longer than 60 seconds to match the old behavior.

My guess is that 25% is a touch conservative, but I think we can figure it out experimentally after we've got everything in.

But anyway, yes, the context should be plumbed through here because it is still worth freeing resources even if it's only for 15 seconds of a queue slot.

Reply:
The filter that imposes the normally-one-minute timeout is earlier (according to request call sequence --- which is opposite of the order of construction) in the handler chain, and does not know anything about the priority and fairness filter (nor should it).  The max-time limit is imposed on the whole rest of the handler chain.  I see no way to break that down into queuing vs executing.  OTOH, we could impose additional limits in our own filter where we _do_ see that structure.  In fact, that is sort of what we do: the priority and fairness filter imposes a limit on queuing time that is shorter than the total limit.  But our filter can not grant extra time for execution; the total limit is imposed without regard to our structure.

Having accepted that we should plumb cancellation into QueueSet::Wait, my remaining question is whether this should be done by passing a cancellation channel or passing the whole Context as an added argument to Wait.